### ISSUE TITLE
fingerprintjs violates the CSP header while trying to set inline-style

### ISSUE THREAD

**Issue by danielstorch** [fingerprintjs/fingerprintjs#733]
### Scenario

- What's the scenario, what happens and what did you expect to happen? 
--> Loading page which uses fingerprintjs but doesn't allow `inline styles`
- What device and browser are you using? 
--> Google Chrome on a Mac
- What version of fingerprintjs are you using?
--> 3.3.2

### Fingerprint data

We have very strict policies on our page and are not allowing `inline-styles`. 
But while using fingerprintjs, the library tries to set some attributes and violates our CSP header. 
Currently i don't understand why fingerprintjs is trying to change something without my knowledge. 

Here is the function of fingerprintjs which causes trouble.
```
function R(e) {
                for (var t = function(e) {
                    for (var t, n, r = "Unexpected syntax '" + e + "'", i = /^\s*([a-z-]*)(.*)$/i.exec(e), o = i[1] || void 0, a = {}, s = /([.:#][\w-]+|\[.+?\])/gi, c = function(e, t) {
                        a[e] = a[e] || [],
                        a[e].push(t)
                    }; ; ) {
                        var u = s.exec(i[2]);
                        if (!u)
                            break;
                        var l = u[0];
                        switch (l[0]) {
                        case ".":
                            c("class", l.slice(1));
                            break;
                        case "#":
                            c("id", l.slice(1));
                            break;
                        case "[":
                            var p = /^\[([\w-]+)([~|^$*]?=("(.*?)"|([\w-]+)))?(\s+[is])?\]$/.exec(l);
                            if (!p)
                                throw new Error(r);
                            c(p[1], null !== (n = null !== (t = p[4]) && void 0 !== t ? t : p[5]) && void 0 !== n ? n : "");
                            break;
                        default:
                            throw new Error(r)
                        }
                    }
                    return [o, a]
                }(e), n = t[0], r = t[1], i = document.createElement(null != n ? n : "div"), o = 0, a = Object.keys(r); o < a.length; o++) {
                    var s = a[o];
                    i.setAttribute(s, r[s].join(" "))
                }
                return i
            }
```

The last line `i.setAttribute(s, r[s].join(" "))` is the issue.
Does someone know why this function is called and what it does and is the only solution to allow `inline style` while using fingerprintjs?

**Comment by Finesse** [comment]
Hi @danielstorch. The code you show is a part of the [dom_blockers](https://github.com/fingerprintjs/fingerprintjs/blob/master/src/sources/dom_blockers.ts) entropy source. The library creates HTML elements with the given attributes (that include `style`) to check which ab blockers are used in the browser, and uses this information to identify the visitor. That is, the library doesn't affect the look of you page, it only modifies the Dom elements created by the library (and removes them after the fingerprint is calculated).

What you can do:

- Remove the `inline style` policy
- Run the library in an iframe with a less strict CSP
- Use the undocumented API to exclude the entropy sources that use inline styles (at your own risk)

**Comment by danielstorch** [comment]
Hey @Finesse, thanks for the answer and explanation. By your proposals you said `Remove the inline style policy`. Did you mean to `add` it?

**Comment by danielstorch** [comment]
Thats perfect, thanks :+1:

**Comment by Finesse** [comment]
The fix will be published in the next release

**Comment by danielstorch** [comment]
@Finesse when do you plan to create a new release with this change?

**Comment by Finesse** [comment]
@danielstorch No plans yet. If you want to use the newest code now, please download the code and [build the code](https://github.com/fingerprintjs/fingerprintjs/blob/master/contributing.md#how-to-build) by yourself.

**Comment by Finesse** [comment]
@danielstorch The change was published in version 3.3.3
